src/jvmMain/kotlin/io/beatmaps/api/collaboration.kt

```diff
@@ -81,17 +85,59 @@ fun Route.collaborationRoute() {
             val req = call.receive<CollaborationResponseData>()
 
             val success = transaction {
-                if (req.accepted) {
-                    Collaboration.update({
-                        Collaboration.id eq req.collaborationId and (Collaboration.collaboratorId eq sess.userId)
-                    }) {
-                        it[accepted] = true
-                    } > 0
-                } else {
-                    Collaboration.deleteWhere {
-                        Collaboration.id eq req.collaborationId and (Collaboration.collaboratorId eq sess.userId)
-                    } > 0
+                val mapId =
+                    if (req.accepted) {
+                        Collaboration.updateReturning(
+                            {
+                                Collaboration.id eq req.collaborationId and (Collaboration.collaboratorId eq sess.userId)
+                            },
+                            {
+                                it[accepted] = true
+                            },
+                            Collaboration.mapId
+                        )
+                    } else {
+                        Collaboration.deleteWhere {
+                            Collaboration.id eq req.collaborationId and (Collaboration.collaboratorId eq sess.userId)
+                        }
+
+                        null
+                    }.let {
+                        it?.firstOrNull()
+                    }
+
+                if (mapId != null) {
+                    // Generate alert for followers of the collaborator.
+                    val map = mapId.let {
+                        BeatmapDao.wrapRow(Beatmap.select {
+                            Beatmap.id eq it[Collaboration.mapId].value
+                        }.single())
+                    }
+
+                    val recipients = Follows.select {
+                        Follows.userId eq sess.userId and Follows.upload and Follows.following
+                    }.map { row ->
+                        row[Follows.followerId].value
+                    }
+
+                    val nonRecipients = Follows.select {
+                        Follows.userId eq map.uploaderId and Follows.following
+                    }.map { row ->
+                        row[Follows.followerId].value
+                    } + listOf(map.uploaderId.value)
+
+                    val finalRecipients = recipients - nonRecipients.toSet()
```

Comment:
Playing some more golf
It's somewhat complex but the database can quite efficiently do this filtering.

```kt
val followsAlias = Follows.alias("f2")
val recipients = Follows
    .join(followsAlias, JoinType.LEFT, followsAlias[Follows.followerId], Follows.followerId) {
        (followsAlias[Follows.userId] eq map.uploaderId) and followsAlias[Follows.following]
    }.select {
        followsAlias[Follows.id].isNull() and (Follows.followerId neq uploaderId) and
            (Follows.userId eq sess.userId) and Follows.upload and Follows.following
    }.map { row ->
        row[Follows.followerId].value
    }.toSet()
```

Reply:
As per discussion on discord these should at least get `.slice()`'d